pypicloud/storage/s3.py

```diff
@@ -151,3 +157,47 @@ def open(self, package):
             yield handle
         finally:
             handle.close()
+
+
+class CloudFrontS3Storage(S3Storage):
+
+    """ Storage backend that uses S3 and CloudFront """
+    def __init__(self, request=None, bucket=None, expire_after=None, bucket_prefix=None,
+                 prepend_hash=None, cloud_front_domain=None, cloud_front_key_file=None,
+                 cloud_front_key_string=None, cloud_front_key_id=None, **kwargs):
+        super(CloudFrontS3Storage, self).__init__(request, bucket, expire_after, bucket_prefix, prepend_hash, **kwargs)
+        self.cloud_front_domain = cloud_front_domain
+        self.cloud_front_key_file = cloud_front_key_file
+        self.cloud_front_key_id = cloud_front_key_id
+        self.cloud_front_key_string = cloud_front_key_string
+
+        self.distribution = Distribution()
+
+    @classmethod
+    def configure(cls, settings):
+        kwargs = super(CloudFrontS3Storage, cls).configure(settings)
+        kwargs['cloud_front_domain'] = getdefaults(
+            settings, 'storage.cloud_front_domain', 'aws.cloud_front_domain', '')
+        kwargs['cloud_front_key_file'] = getdefaults(
+            settings, 'storage.cloud_front_key_file', 'aws.cloud_front_key_file', None)
+        kwargs['cloud_front_key_string'] = getdefaults(
+            settings, 'storage.cloud_front_key_string', 'aws.cloud_front_key_string', None)
+        kwargs['cloud_front_key_id'] = getdefaults(
+            settings, 'storage.cloud_front_key_id', 'aws.cloud_front_key_id', '')
+
+        return kwargs
+
+    def get_url(self, package):
+        """ Get the fully-qualified CloudFront path for a package """
+        path = self.calculate_path(package)
+        url = self.cloud_front_domain + '/' + path
+
+        url = url.replace('+', '%2b')
```

Comment:
The problem is that quote replaces stuff that shouldn't be replaced:

    >>> quote('https://domain.com/test/api/?param=val')
    'https%3A//domain.com/test/api/%3Fparam%3Dval'

I tried to use quote in the test setup that I have now and the links don't work anymore. I know for sure that the `+` sign needs to be quoted and there might be a few more characters that we need to replace as well, but replacing all of them will break the links.